File: 백준/Gold/1937. 욕심쟁이 판다/욕심쟁이 판다.py
"""
그래프
대나무가 적은곳에서 많은곳으로 간선이 존재한다고 생각

완탐?
2500 * O(V(2500) + E(10000)) 시초

위상정렬?
"""
import sys
from collections import deque

input = lambda: sys.stdin.readline().strip()

dxy = [[-1, 0], [0, 1], [1, 0], [0, -1]]

n = int(input())
lst = []
for _ in range(n):
    lst.append(list(map(int, input().split())))

in_degree = [[0 for _ in range(n)] for _ in range(n)]
distance = [[-1 for _ in range(n)] for _ in range(n)]

q = deque()

# 차수 정리 & 차수가 0인 정점 큐에 삽입
for i in range(n):
    for j in range(n):
        cnt = 0
        for dy, dx in dxy:
            ny = i + dy
            nx = j + dx
            if 0 <= ny < n and 0 <= nx < n:
                if lst[ny][nx] < lst[i][j]:
                    cnt += 1
        in_degree[i][j] = cnt
        if cnt == 0:
            q.append((i, j))
            distance[i][j] = 1

# 큐 돌리기
while q:
    cur_y, cur_x = q.popleft()
    for dy, dx in dxy:
        ny = cur_y + dy
        nx = cur_x + dx
        if 0 <= ny < n and 0 <= nx < n:
            if lst[ny][nx] > lst[cur_y][cur_x]:
                in_degree[ny][nx] -= 1
                if in_degree[ny][nx] == 0:
                    distance[ny][nx] = distance[cur_y][cur_x] + 1
                    q.append((ny, nx))

answer = 0
for i in distance:
    answer = max(answer, max(i))

print(answer)

File: 백준/Gold/1937. 욕심쟁이 판다/욕심쟁이 판다.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();
    private static int n;
    private static int[][] lst, in_degree, distance;
    private static Queue<Integer> q;

    public static void main(String[] args) throws IOException {

        int[][] dxy = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

        n = Integer.parseInt(br.readLine());
        lst = new int[n][n];
        in_degree = new int[n][n];
        distance = new int[n][n];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                lst[i][j] = Integer.parseInt(st.nextToken());
            }
            Arrays.fill(in_degree[i], 0);
            Arrays.fill(distance[i], -1);
        }
        
        q = new ArrayDeque<Integer>();

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < 4; k++) {
                    int ny = i + dxy[k][0];
                    int nx = j + dxy[k][1];
                    if (0 <= ny && ny < n && 0 <= nx && nx < n) {
                        if (lst[i][j] > lst[ny][nx]) {
                            in_degree[i][j] += 1;
                        }
                    }
                }
                if (in_degree[i][j] == 0) {
                    q.offer(i);
                    q.offer(j);
                    distance[i][j] = 1;
                }
            }
        } // end of for i

        int answer = 1;
        while (!q.isEmpty()) {
            int curY = q.poll();
            int curX = q.poll();
            for (int k = 0; k < 4; k++) {
                int ny = curY + dxy[k][0];
                int nx = curX + dxy[k][1];
                if (0 <= ny && ny < n && 0 <= nx && nx < n) {
                    if (lst[ny][nx] > lst[curY][curX]) {
                        in_degree[ny][nx] -= 1;
                        if (in_degree[ny][nx] == 0) {
                            distance[ny][nx] = distance[curY][curX] + 1;
                            q.offer(ny);
                            q.offer(nx);
                            answer = Math.max(answer, distance[ny][nx]);
                        }
                    }
                }
            }
        } // end of while

        sb.append(answer);
        System.out.println(sb);
        
    } // end of main
} // end of class